File: CardDeck.py
from enum import Enum
from io import StringIO
from random import randint, shuffle
from typing import List


class StringBuffer(StringIO):

    def __init__(self):
        super().__init__()

    def write(self, __s: str) -> int:
        pos = self.tell()
        self.seek(len(self.getvalue()))
        super().write(__s)
        self.seek(pos)

    def writeline(self, line: str):
        self.write(f"{line}\n")


class Color(Enum):
    EICHEL = 0
    SCHELLE = 1
    HERZ = 2
    BLATT = 3

    def __gt__(self, other):
        if not isinstance(other, Color):
            raise ValueError
        return self.value > other.value

    def __lt__(self, other):
        if not isinstance(other, Color):
            raise ValueError
        return self.value < other.value

    def __eq__(self, other):
        if not isinstance(other, Color):
            raise ValueError
        return self.value == other.value

    def __ne__(self, other):
        if not isinstance(other, Color):
            raise ValueError
        return self.value != other.value


class Value(Enum):
    SIX = 6
    SEVEN = 7
    EIGHT = 8
    NINE = 9
    TEN = 10
    BUBE = 11
    DAME = 12
    KÖNIG = 13
    ASS = 14

    def __gt__(self, other):
        if not isinstance(other, Value):
            raise ValueError
        return self.value > other.value

    def __lt__(self, other):
        if not isinstance(other, Value):
            raise ValueError
        return self.value < other.value

    def __eq__(self, other):
        if not isinstance(other, Value):
            raise ValueError
        return self.value == other.value

    def __ne__(self, other):
        if not isinstance(other, Value):
            raise ValueError
        return self.value != other.value


class Card:
    color: Color
    value: Value

    def __init__(self, color: Color, value: Value) -> None:
        self.color = color
        self.value = value

    def __str__(self) -> str:
        return self.__repr__()

    def __repr__(self) -> str:
        return f"[{self.color}, {self.value}]"

    def getColor(self):
        return self.color

    def getValue(self):
        return self.value

    def __eq__(self, other):
        if not isinstance(other, Card):
            raise ValueError
        return self.value == other.value and self.color == other.color


class Player:

    cards: List[Card]
    name: str
    stdout: StringBuffer
    stdin: StringBuffer

    def __init__(self, name: str):
        self.name = name
        self.stdout = StringBuffer()
        self.stdin = StringBuffer()
        self.cards = []

    def addCard(self, card: Card):
        self.cards.append(card)

    def getCard(self, index: int) -> Card:
        return self.cards.pop(index)

    def listCards(self) -> List[Card]:
        return self.cards

    def writeOutput(self, message: str):
        """sends the player a message"""
        self.stdout.writeline(message)

    def readInput(self, length: int = -1) -> str:
        """reads wether there has been a textinput"""
        return self.stdin.readline(length)

    def awaitInput(self, validation) -> str:
        """Reads the user input until validation returns something"""
        while True:
            try:
                text = self.readInput()
                if len(text) > 0:
                    return validation(text)
            except:
                self.writeOutput("Invalid Input")
                continue


class Deck:

    cards: List[Card] = []

    def __init__(self, ignored_values: List[Value]):
        for color in Color:
            for value in Value:
                if value not in ignored_values:
                    self.cards.append(Card(color, value))

    def shuffle(self):
        shuffle(self.cards)

    def lift(self):
        """= Abheben """
        index = randint(0, len(self.cards))
        self.cards = self.cards[index:len(self.cards)] + self.cards[0:index]

    def peek(self, index: int = -1) -> Card:
        return self.cards[index]

    def take(self, index: int = 0) -> Card:
        return self.cards.pop(index)

    def __str__(self) -> str:
        return self.__repr__()

    def __repr__(self) -> str:
        return f"[{', '.join(str(card) for card in self.cards)}]"


if __name__ == "__main__":
    d = Deck([Value.SIX])
    d.shuffle()
    d.lift()
    print(d)
    print(d.peek())

File: CardDeck.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

type Color int8
type Value int8

const (
	EICHEL Color = iota
	SCHELLE
	HERZ
	BLATT
	SIX   Value = 6
	SEVEN Value = 7
	EIGHT Value = 8
	NINE  Value = 9
	TEN   Value = 10
	BUBE  Value = 11
	DAME  Value = 12
	KÖNIG Value = 13
	ASS   Value = 14
)

func (color Color) String() string {
	switch color {
	case EICHEL:
		return "Eichel"
	case SCHELLE:
		return "Schelle"
	case HERZ:
		return "Herz"
	case BLATT:
		return "Blatt"
	}
	return "" // unreacheable
}

func (value Value) String() string {
	switch value {
	case SIX:
		return "Sechs"
	case SEVEN:
		return "Sieben"
	case EIGHT:
		return "Acht"
	case NINE:
		return "Neun"
	case TEN:
		return "Zehn"
	case BUBE:
		return "Bube"
	case DAME:
		return "Dame"
	case KÖNIG:
		return "König"
	case ASS:
		return "Ass"
	}
	return ""
}

type Card struct {
	color Color
	value Value
}

func (card Card) getColor() Color {
	return card.color
}

func (card Card) getValue() Value {
	return card.value
}

func (card Card) String() string {
	return "[" + card.color.String() + " " + card.value.String() + "]"
}

func listToString(cards []Card) string {
	str := "["
	for index, card := range cards {
		if index > 0 {
			str += ", "
		}
		str += card.String()
	}
	return str + "]"
}

type Player struct {
	cards  []Card
	name   string
	stdout chan string
	stdin  chan string
}

func (player Player) String() string {
	return player.name + ": " + listToString(player.cards)
}

func (player *Player) addCard(card Card) {
	player.cards = append(player.cards, card)
}

func (player *Player) getCard(index int8) Card {
	card := player.cards[index]
	shifted := player.cards[index+1:]
	player.cards = append(player.cards[:index], shifted...)
	return card
}

type Deck struct {
	cards []Card
}

func createDeck() Deck {
	deck := Deck{[]Card{}}
	for color := EICHEL; color <= BLATT; color++ {
		for value := SIX; value <= ASS; value++ {
			deck.cards = append(deck.cards, Card{color, value})
		}
	}
	return deck
}

func (deck Deck) shuffle() {
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(deck.cards), func(i, j int) { deck.cards[i], deck.cards[j] = deck.cards[j], deck.cards[i] })
}

func (deck *Deck) lift() {
	index := rand.Intn(len(deck.cards) - 1)
	lifted := deck.cards[:index]
	deck.cards = append(deck.cards[index:len(deck.cards)], lifted...)
}

func (deck Deck) peek(index int) Card {
	if index == -1 {
		index = len(deck.cards) - 1
	}
	return deck.cards[index]
}

func (deck *Deck) take(index int) Card {
	if index == -1 {
		index = len(deck.cards) - 1
	}
	card := deck.cards[index]
	shifted := deck.cards[index+1:]
	deck.cards = append(deck.cards[:index], shifted...)
	return card
}

func (deck Deck) String() string {
	return listToString(deck.cards)
}

func main() {
	fmt.Println(ASS > KÖNIG)
	card := Card{EICHEL, ASS}
	fmt.Println(card.getColor(), card.getValue())

	player := Player{[]Card{card}, "somename", make(chan string), make(chan string)}
	player.addCard(Card{HERZ, ASS})
	fmt.Println(listToString(player.cards))
	fmt.Println(player.getCard(0))
	fmt.Println(listToString(player.cards))

	deck := createDeck()
	deck.shuffle()
	deck.lift()
	fmt.Println(deck.peek(-1))
	fmt.Println(deck.peek(-1) == deck.take(-1))
}
